File: js/test_money.js
const assert = require('assert');


class Money{
  constructor(amount,currency){
    this.amount = amount;
    this.currency = currency;
  }

  times(multiplier){
    return new Money(this.amount * multiplier, this.currency)
  }
  divide(divisor){
    return new Money(this.amount / divisor, this.currency)
  }
}

let fiveDollars = new Money(5, "USD");
let tenDollars = new Money(10, "USD");
assert.deepStrictEqual(fiveDollars.times(2), tenDollars, 'Expected 10 dollars');

let tenEuros = new Money(10,"EUR")
let twentyEuros = new Money(20,"EUR")
assert.deepStrictEqual(tenEuros.times(2), twentyEuros, 'Expected 20 euros');

let originalMoney = new Money(4002.0, "KRW");
let actualMoneyAfterDivision = originalMoney.divide(4);
let expectedMoneyAfterDivision = new Money(1000.5, "KRW");
assert.deepStrictEqual(actualMoneyAfterDivision, expectedMoneyAfterDivision,
  `Expected: ${JSON.stringify(expectedMoneyAfterDivision)}, got: ${JSON.stringify(actualMoneyAfterDivision)}`);

File: py/test_money.py
import unittest
from money import Money
from portfolio import Portfolio
    # return Money(total_amount, currency) 

class TestMoney(unittest.TestCase):
  def testMultiplication(self):
    tenEuros = Money(10,"EUR")
    twentyEuros = Money(20,"EUR") 
    self.assertEqual(tenEuros.times(2), twentyEuros)

  def testDivision(self):
    originalMoney = Money(1000.5, "KRW")
    actualMoneyAfterDivision = originalMoney.divide(4)
    expectedMoneyAfterDivision = Money(250.125, "KRW")

    self.assertEqual(expectedMoneyAfterDivision, actualMoneyAfterDivision)
  
  def testAddition(self):
    fiveDollars = Money(5, "USD")
    tenDollars = Money(10, "USD")
    fifteenDollars = Money(15, "USD")
    
    portfolio = Portfolio()
    portfolio.add(fiveDollars, tenDollars)
    
    self.assertEqual(portfolio.evaluate("USD"), fifteenDollars, 'Expected portfolio to contain 5 and 10 dollars')
    
  def testAdditionOfDollarsAndEuros(self):
    fiveDollars = Money(5, "USD")
    tenEuros = Money(10, "EUR")
    
    portfolio = Portfolio()
    portfolio.add(fiveDollars, tenEuros)
    
    # Assuming the conversion rate is 1 EUR = 1.2 USD
    expectedValue= Money(17, "USD") 
    actualValue = portfolio.evaluate("USD")
    
    self.assertEqual(expectedValue,actualValue, '%s != %s' % (expectedValue, actualValue))
  
  def testAdditionOfDollarsAndWons(self):
    """Test addition of two different currencies: Dollars and Wons"""
    oneDollary = Money(1, "USD")
    elevenHundredWon = Money(1100, "KRW")
    
    portfolio = Portfolio()
    portfolio.add(oneDollary, elevenHundredWon)
    
    # Assuming the conversion rate is 1 USD = 1000 KRW for simplicity
    expectedValue = Money(2200, "KRW") 
    actualValue = portfolio.evaluate("KRW")
    
    self.assertEqual(expectedValue, actualValue, '%s != %s' % (expectedValue, actualValue))
  
  def testAddtionWithMultipleMissingExchangeRates(self):
    """Test addition with multiple missing exchange rates"""
    oneDollar = Money(1, "USD")
    oneEuro = Money(1, "EUR")
    oneWon = Money(1, "KRW")
    
    portfolio = Portfolio()
    portfolio.add(oneDollar, oneEuro, oneWon)
    
    # Assuming no exchange rates are defined for these currencies
    expectedErrorMessage = "Missing exchange rate(s):[USD->kalgnid,EUR->kalgnid,KRW->kalgnid]"
    
    with self.assertRaises(Exception) as context:
      portfolio.evaluate("kalgnid")
    
    # self.assertEqual(str(context.exception), expectedErrorMessage)
if __name__ == '__main__':
    unittest.main()
